crates/hir-ty/src/infer/cast.rs

```diff
@@ -1,47 +1,392 @@
+                            TyKind::Scalar(Scalar::Int(_) | Scalar::Uint(_) | Scalar::Float(_))
+                            | TyKind::InferenceVar(
+                                _,
+                                TyVariableKind::Integer | TyVariableKind::Float,
+                            ) => Err(CastError::NeedDeref),
+
+                            _ => Err(CastError::NeedViaPtr),
+                        },
+                        // array-ptr-cast
+                        CastTy::Ptr(t, m) => {
+                            let t = table.resolve_ty_shallow(&t);
+                            if !is_sized(&t, table) {
+                                return Err(CastError::IllegalCast);
+                            }
+                            check_ref_cast(&inner_ty, *mutbl, &t, m, table).map(|_| adjustments)
+                        }
+                        _ => Err(CastError::NonScalar),
+                    };
+                }
+                _ => return Err(CastError::NonScalar),
+            },
+            _ => return Err(CastError::NonScalar),
+        };
+
+    // rustc checks whether the `expr_ty` is foreign adt with `non_exhaustive` sym
+
+    match (t_from, t_cast) {
+        (_, CastTy::Int(Int::CEnum) | CastTy::FnPtr) => Err(CastError::NonScalar),
+        (_, CastTy::Int(Int::Bool)) => Err(CastError::CastToBool),
+        (CastTy::Int(Int::U(UintTy::U8)), CastTy::Int(Int::Char)) => Ok(adjustments),
+        (_, CastTy::Int(Int::Char)) => Err(CastError::CastToChar),
+        (CastTy::Int(Int::Bool | Int::CEnum | Int::Char), CastTy::Float) => {
+            Err(CastError::NeedViaInt)
+        }
+        (CastTy::Int(Int::Bool | Int::CEnum | Int::Char) | CastTy::Float, CastTy::Ptr(..))
+        | (CastTy::Ptr(..) | CastTy::FnPtr, CastTy::Float) => Err(CastError::IllegalCast),
+        (CastTy::Ptr(src, _), CastTy::Ptr(dst, _)) => {
+            check_ptr_ptr_cast(&src, &dst, table).map(|_| adjustments)
+        }
+        (CastTy::Ptr(src, _), CastTy::Int(_)) => {
+            check_ptr_addr_cast(&src, table).map(|_| adjustments)
+        }
+        (CastTy::Int(_), CastTy::Ptr(dst, _)) => {
+            check_addr_ptr_cast(&dst, table).map(|_| adjustments)
+        }
+        (CastTy::FnPtr, CastTy::Ptr(dst, _)) => {
+            check_fptr_ptr_cast(&dst, table).map(|_| adjustments)
+        }
+        (CastTy::Int(Int::CEnum), CastTy::Int(_)) => Ok(adjustments),
+        (CastTy::Int(Int::Char | Int::Bool), CastTy::Int(_)) => Ok(adjustments),
+        (CastTy::Int(_) | CastTy::Float, CastTy::Int(_) | CastTy::Float) => Ok(adjustments),
+        (CastTy::FnPtr, CastTy::Int(_)) => Ok(adjustments),
+    }
+}
+
+fn check_ref_cast(
+    t_expr: &Ty,
+    m_expr: Mutability,
+    t_cast: &Ty,
+    m_cast: Mutability,
+    table: &mut InferenceTable<'_>,
+) -> Result<(), CastError> {
+    // Mutability order is opposite tu rustc. `Mut < Not`
```

Comment:
Might be good to change that on our side then (separately from this), the less we diverge from rustc with things the better we can draw comparisons between the code bases (which imo for type check is quite important)

Reply:
I agree. But this type is from `chalk_ir`. Maybe we should change this at there?

Reply:
BTW, another typo "tu" here 😅

Reply:
Ah I see, probably no point in changing it then. But we should keep that in mind for the trait solver migration 😬

Reply:
Yes, there are many places in our code that tailored for chalk quirks and I can remember some of them. The ones that no one remembers or knows would be covered by our tests, I hope 😨